homework/index.js

```diff
@@ -1,46 +1,169 @@
 'use strict';
 
 {
-  function fetchJSON(url, cb) {
-    const xhr = new XMLHttpRequest();
-    xhr.open('GET', url);
-    xhr.responseType = 'json';
-    xhr.onload = () => {
-      if (xhr.status >= 200 && xhr.status <= 299) {
-        cb(null, xhr.response);
-      } else {
-        cb(new Error(`Network error: ${xhr.status} - ${xhr.statusText}`));
-      }
-    };
-    xhr.onerror = () => cb(new Error('Network request failed'));
-    xhr.send();
-  }
-
   function createAndAppend(name, parent, options = {}) {
     const elem = document.createElement(name);
     parent.appendChild(elem);
     Object.keys(options).forEach(key => {
       const value = options[key];
       if (key === 'text') {
         elem.textContent = value;
+      } else if (key === 'html') {
+        elem.innerHTML = value;
       } else {
         elem.setAttribute(key, value);
       }
     });
     return elem;
   }
 
-  function main(url) {
-    fetchJSON(url, (err, repositories) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
-        return;
-      }
-      createAndAppend('pre', root, { text: JSON.stringify(repositories, null, 2) });
+  function addRow(tbody, label, value) {
+    const tr = createAndAppend('tr', tbody);
+    createAndAppend('td', tr, { text: `${label}:` });
+    createAndAppend('td', tr, { text: value });
+    return tr;
+  }
+
+  function renderError(error, container) {
+    createAndAppend('div', container, { text: error.message, class: 'alert-error' });
+  }
+
+  async function bringData(url) {
+    const response = await fetch(url);
+    if (!response.ok) {
+      throw Error(`Network error: ${response.status} - ${response.statusText}`);
+    } else if (response.statusText !== 'OK') {
+      return null;
+    } else {
+      return response.json();
+    }
+  }
+
+  function renderContributions(contributors, infoRight) {
+    if (contributors === null) {
+      createAndAppend('p', infoRight, {
+        class: 'no-content',
+        html: 'This repository is empty.<br /> It has No content. <br /> It has No contributions.',
+      });
+      return;
+    }
+    if (contributors.length === 0) {
+      createAndAppend('p', infoRight, {
+        class: 'no-content',
+        text: 'There are no contributions yet.',
+      });
+      return;
+    }
+
+    contributors.forEach(contributor => {
```

Comment:
The contributor's list should be semantically expressed as an HTML unordered list (`ul`). This, for the benefit of visually impaired users that use a screen reader to narrate what is on the screen (ARIA-compliance). Your code generates a containing `div` with each contributor represent by an `a` tag:

```html
<div class="info-right">
  Contributions:
  <a class="contributor-link" href="https://github.com/gijscor" target="_blank">
    <div class="contributor-img-div">
      <img
        class="contributor-img"
        src="https://avatars0.githubusercontent.com/u/15912395?v=4"
        alt="Contributor photo"
        style="max-height: 50px;"
      />
    </div>
    <div class="contributor-data-div">
      <div class="contributor-name">gijscor</div>
      <div class="contributor-contributions">47</div>
    </div></a
  >
  <hr class="hr-contributor" />
  <a class="contributor-link" href="https://github.com/wouterkleijn" target="_blank">
    <div class="contributor-img-div">
      <img
        class="contributor-img"
        src="https://avatars1.githubusercontent.com/u/37488847?v=4"
        alt="Contributor photo"
        style="max-height: 50px;"
      />
    </div>
    <div class="contributor-data-div">
      <div class="contributor-name">wouterkleijn</div>
      <div class="contributor-contributions">11</div>
    </div></a
  >
  <hr class="hr-contributor" />
</div>
```

A structure like this would better meet the ARIA-compliance:

```html

<div>
  <p> Contributions:</p>
  <ul>
    <li>
      <a href="...">
        <!-- contributor details-->
      </a>
    </li>
  </ul>
</div>
```

Reply:
yes that is very important point, I have to consider it very good in my work.